contracts/utils/Swap.sol

```diff
@@ -0,0 +1,487 @@
+    }
+
+    /**
+     * @notice Cancel one or more nonces
+     * @dev Cancelled nonces are marked as used
+     * @dev Emits a Cancel event
+     * @dev Out of gas may occur in arrays of length > 400
+     * @param nonces uint256[] List of nonces to cancel
+     */
+    function cancelNonce(uint256[] calldata nonces) external override {
+        for (uint256 i = 0; i < nonces.length; i++) {
+            uint256 nonce = nonces[i];
+            if (_markNonceAsUsed(msg.sender, nonce)) {
+                emit Cancel(nonce, msg.sender);
+            }
+        }
+    }
+
+    /************************************************
+     *  PUBLIC VIEW FUNCTIONS
+     ***********************************************/
+
+    /**
+     * @notice Validates Swap bid for any potential errors
+     * @param bid Bid struct containing bid details
+     * @return tuple of error count and bytes32[] memory array of error messages
+     */
+    function check(Bid calldata bid)
+        public
+        view
+        override
+        returns (uint256, bytes32[] memory)
+    {
+        Offer memory offer = swapOffers[bid.swapId];
+        require(offer.seller != address(0), "Offer does not exist");
+
+        bytes32[] memory errors = new bytes32[](MAX_ERROR_COUNT);
+
+        uint256 errCount;
+
+        // Check signature
+        address signatory = _getSignatory(bid);
+
+        if (signatory == address(0)) {
+            errors[errCount] = "SIGNATURE_INVALID";
+            errCount++;
+        }
+
+        if (signatory != bid.signerWallet) {
+            errors[errCount] = "SIGNATURE_MISMATCHED";
+            errCount++;
+        }
+
+        // Check nonce
+        if (nonceUsed(signatory, bid.nonce)) {
+            errors[errCount] = "NONCE_ALREADY_USED";
+            errCount++;
+        }
+
+        // Check bid size
+        if (bid.buyAmount < offer.minBidSize) {
+            errors[errCount] = "BID_TOO_SMALL";
+            errCount++;
+        }
+        if (bid.buyAmount > offer.availableSize) {
+            errors[errCount] = "BID_EXCEED_AVAILABLE_SIZE";
+            errCount++;
+        }
+
+        // Check bid price
+        uint256 bidPrice =
+            (uint256(bid.sellAmount) * uint256(10)**OTOKEN_DECIMALS) /
+                bid.buyAmount;
+        if (bidPrice < offer.minPrice) {
+            errors[errCount] = "PRICE_TOO_LOW";
+            errCount++;
+        }
+
+        // Check signer allowance
+        uint256 signerAllowance =
+            IERC20(offer.biddingToken).allowance(
+                bid.signerWallet,
+                address(this)
+            );
+        if (signerAllowance < bid.sellAmount) {
+            errors[errCount] = "SIGNER_ALLOWANCE_LOW";
+            errCount++;
+        }
+
+        // Check signer balance
+        uint256 signerBalance =
+            IERC20(offer.biddingToken).balanceOf(bid.signerWallet);
+        if (signerBalance < bid.sellAmount) {
+            errors[errCount] = "SIGNER_BALANCE_LOW";
```

Comment:
We will also have to check the `seller`'s otoken balance to see if there is sufficient balance to fill the bid